### PR TITLE
Revamp UI tests

### LINKED ISSUE #1644: Enable UI testing for click_tests
click tests don't currently support UI testing.

The UI test is set up within the `client` fixture, we would need to do the same setup for the `device_handler` fixture, and maybe make sure this works across restarts?

### LINKED ISSUE #2494: tests: fixtures.suggestions omit other models
it should be possible to simply `cp fixtures.suggestion.json fixtures.json` to apply the suggestion. This breaks because the suggestions file does not include results from other models.

### PR DESCRIPTION
**Pull request by matejcik**
* this PR enables UI tests for `click_tests`, which fixes #1644
* UI diffs are now smarter and can figure out that a screen was added or removed in the middle. This works through python's built in `difflib`.
* Option `--not-generate-report-after-each-test` was removed, this behavior is now the default. Instead, `show_results.py` will regenerate the index if you ask for it.
* `fixtures.diff` has been removed in favor of using per-testcase metadata, and being able to collect tests from last run via `TestResult.recent_tests()`. Use `./tests/update_fixtures.py` to apply any changes.
* fixes #2494
* In HTML reports, images are now placed as separate resources instead of inline base64. This makes the html pages (esp. the big ones) smaller and more manageable, and uses less space/bandwidth because images are deduplicated.
* the system is now overall smarter in figuring out which models to touch
* got rid of most of the globals in the UI test module
* re-enabled autolock passphrase test (#922)

### CODE REVIEW

**Review by grdddj — COMMENTED** [review]
Quick review, without going too much into the code itself:

- there are no click tests for BIP39 recovery, only for SLIP39 - we should add them to check the keyboard
- new reports do not load correctly for me, a lot of the pictures are missing, the later the test, the less pictures are there (tried multiple reports in two browsers)
--- e.g. https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3690613386/artifacts/test_ui_report/all_screens.html
--- e.g. https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3690613403/artifacts/test_ui_report/all_unique_screens.html
image
or 
image


- `make test_emu_ui_multicore` do not work currently (`FileNotFoundError: [Errno 2] No such file or directory: '/home/jmusil/trezor-firmware/tests/ui_tests/reports/test/images/bd83a31d0fc4c23953dfd0d138e4441984e34698ace96aad5308a4ae51b712ae.png'`) - no images are saved there
- what is the purpose of `tests/ui_tests/reports/.keep`?
- how to update all the files/reports in `trezor-firmware/tests/ui_tests/reporting/reports/test`? I like to access the `file:///home/jmusil/trezor-firmware/tests/ui_tests/reporting/reports/test` and see everything there
--- `../tests/show_results.py` opens up the browser with the main report, but the `index.html` is not changed
- why not having the `fixtures.json` indented? Could improve the one-glance readability of the data structure there

**Comment by matejcik** [comment]
> new reports do not load correctly for me, a lot of the pictures are missing, the later the test, the less pictures are there (tried multiple reports in two browsers)

well shoot, you are right. gitlab is rate-limiting us with 429 too many requests :(
this sucks because as I was saying previously, the documents with embedded images don't work on my side 80 % of the time and I was hoping for this to be a solution. wondering if we could have our cake and eat it too somehow -- client-side throttling for the images maybe?

**Comment by matejcik** [comment]
> * there are no click tests for BIP39 recovery, only for SLIP39 - we should add them to check the keyboard

yes, let's make an issue for that.

> * what is the purpose of `tests/ui_tests/reports/.keep`?

makes sure the `reports` subdir is part of the git checkout.
honestly not sure if we actually need it. it was there before, it's moved from "reporting/reports" to "reports"

> * how to update all the files/reports in `trezor-firmware/tests/ui_tests/reporting/reports/test`? I like to access the `file:///home/jmusil/trezor-firmware/tests/ui_tests/reporting/reports/test` and see everything there
>   --- `../tests/show_results.py` opens up the browser with the main report, but the `index.html` is not changed

not sure what you mean. for me when I refresh the main page, through the server, the index.html gets regenerated.
the headers are set correctly for no caching so every reload should force the browser to re-query, which regenerates the index

> * why not having the `fixtures.json` indented? Could improve the one-glance readability of the data structure there

fixtures.json is too long to read at a glance anyway? :woman_shrugging: the indents will tell you nothing when viewing a diff, so it looks to me it would be just a waste of four spaces per line

**Comment by grdddj** [comment]
> > * there are no click tests for BIP39 recovery, only for SLIP39 - we should add them to check the keyboard
> 
> yes, let's make an issue for that.

Made https://github.com/trezor/trezor-firmware/issues/2781 

> > * how to update all the files/reports in `trezor-firmware/tests/ui_tests/reporting/reports/test`? I like to access the `file:///home/jmusil/trezor-firmware/tests/ui_tests/reporting/reports/test` and see everything there
> >   --- `../tests/show_results.py` opens up the browser with the main report, but the `index.html` is not changed
> 
> not sure what you mean. for me when I refresh the main page, through the server, the index.html gets regenerated. the headers are set correctly for no caching so every reload should force the browser to re-query, which regenerates the index

I mean this directory I was using to look for results after the tests finished in all previous branches. It is not being updated currently.

image

File: tests/ui_tests/common.py [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+from typing_extensions import Self
+
+from trezorlib.debuglink import TrezorClientDebugLink as Client
+
+UI_TESTS_DIR = Path(__file__).resolve().parent
+SCREENS_DIR = UI_TESTS_DIR / "screens"
+IMAGES_DIR = SCREENS_DIR / "all_images"
+FIXTURES_FILE = UI_TESTS_DIR / "fixtures.json"
+
+FixturesType = t.NewType("FixturesType", "dict[str, dict[str, dict[str, str]]]")
```
[the remark this answers is not in the document]
Comment (matejcik) [reply tests/ui_tests/common.py:35]: f846ef8e3d

**Comment by matejcik** [comment]
lazy loading seems to help a lot for me. i'm still getting some 429s but due to caching behavior I can just reload the page and the missing images will come in fine

**Review by grdddj — COMMENTED** [review]
Superb! The lazyloading is working fine for me:
- https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3697637816/artifacts/test_ui_report/all_screens.html
- https://satoshilabs.gitlab.io/-/trezor/trezor-firmware/-/jobs/3697637816/artifacts/test_ui_report/all_unique_screens.html

`make test_emu_multicore` works for me as well

I must still do something wrong, because from the reports, only `index.html` is being updated (after running `../tests/show_results.py`) . Seems almost like `testreport.generate_reports()` is not being called.

image

**Comment by grdddj** [comment]
On a general note, we can merge this before https://github.com/trezor/trezor-firmware/pull/2750 and integrate current changes into that later (if any, plus probably the diff-searching algorithm). Then there is also https://github.com/trezor/trezor-firmware/pull/2699 which I am not really sure whether is still applicable - @TychoVrahe?

After this PR is merged, I want to do click-tests for `TR`, it would be nice to have the UI recording ready.

**Comment by grdddj** [comment]
> I must still do something wrong, because from the reports, only `index.html` is being updated (after running `../tests/show_results.py`) . Seems almost like `testreport.generate_reports()` is not being called.

Yes, I was doing something wrong, I tried it without having `--ui=test`. Sorry for false alarm, works now

**Review by grdddj — APPROVED** [review]

**Comment by TychoVrahe** [comment]
#2699 probably partially still relevant - the part that fixes OK&UPDATE on the last failed test. unless that was removed or fixed here somehow, didn't check but seems out of scope for this

**Comment by matejcik** [comment]
added the javascript fix in d5dbdaa.
the test generation thing seems fixed as part of the other changes

**Pull request merged**